cli/azd/pkg/tools/azcli/ad.go

```diff
@@ -367,3 +372,68 @@ func (cli *azCli) createRoleAssignmentsClient(
 
 	return client, nil
 }
+
+// Get role definition id from role assignment within specific scope and principalID
+func (cli *azCli) getUserRoleDefinitionIdList(
+	ctx context.Context,
+	subscriptionId string,
+	scope string,
+	principalId string,
+) ([]string, error) {
+	client, err := cli.createRoleAssignmentsClient(ctx, subscriptionId)
+	if err != nil {
+		return nil, err
+	}
+
+	pager := client.NewListForScopePager(scope,
+		&armauthorization.RoleAssignmentsClientListForScopeOptions{
+			Filter: convert.RefOf(fmt.Sprintf("atScope() and assignedTo('%s')", principalId)),
+		})
+
+	roleDefinitionIds := []string{}
+
+	for pager.More() {
+		page, err := pager.NextPage(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("failed getting next page of role assignments on scope '%s': %w", scope, err)
+		}
+		for _, v := range page.Value {
+			roleDefinitionIds = append(roleDefinitionIds, *v.Properties.RoleDefinitionID)
+		}
+	}
+
+	if len(roleDefinitionIds) == 0 {
+		return nil, fmt.Errorf("role assignment with scope '%s' was not found", scope)
+	}
+
+	return roleDefinitionIds, nil
+}
+
+// Get role definition name from role definition id
+func (cli *azCli) GetUserRoleDefinitionName(
+	ctx context.Context,
+	subscriptionId string,
+	scope string,
+	principalId string,
+) ([]string, error) {
+	roleDefinitionName := []string{}
+	client, err := cli.createRoleDefinitionsClient(ctx, subscriptionId)
+	if err != nil {
+		return nil, err
+	}
+
+	roleDefinitionID, err := cli.getUserRoleDefinitionIdList(ctx, subscriptionId, scope, principalId)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, id := range roleDefinitionID {
+		roleDefinitionResult, err := client.GetByID(ctx, id, nil)
+		if err != nil {
+			return nil, fmt.Errorf("failing to get role definition name from role definition id: %w", err)
+		}
+		roleDefinitionName = append(roleDefinitionName, *roleDefinitionResult.Properties.RoleName)
```

Comment:
Can a role definition have an optional name? If so, `RoleName` may be `nil` here and requires handling.